src/sparseml/exporters/transforms/kv_cache/cache_keys_and_values.py

```diff
@@ -356,19 +386,50 @@ def _find_key_value_matmul_pairs(
     return key_value_matmul_pairs
 
 
-def _is_value_matmul(node: NodeProto, graph: ONNXGraph) -> bool:
-    # A valid value MatMul needs to meet the following criteria:
-    #   - is_matmul(node) is True
-    #   - have no parameters
-    #   - have a single parent node: a Softmax
+def is_value_matmul(
+    node: NodeProto,
+    graph: ONNXGraph,
+    allowed_nodes_before_softmax: Set[str] = ALLOWED_NODES_BEFORE_SOFTMAX,
+) -> bool:
+    """
+    Returns True if the node is a "value" MatMul, i.e. a MatMul that meets
+    the following criteria:
+        -   is_matmul(node) is True
+        -   node has no parameters
+        -   node has a single `Softmax` parent node
+                or
+            the parent node `Softmax` is preceded by
+            a set of nodes that are specified in the
+            `allowed_nodes_before_softmax` set
+
+    :param node: node to check
+    :param graph: graph containing the node
+    :param allowed_nodes_before_softmax: set of node types that are allowed
+        to be located between the node in question a Softmax node, so that
+        the node can still be considered a "value" MatMul
+    """
 
     if not is_matmul(node) or _is_parameterized_matmul(node, graph):
         # not a matmul or MatMul op has a parameter
         return False
 
     for parent in graph.get_node_parents(node):
-        if not isinstance(parent, NodeProto):
-            continue
+        while parent.op_type in allowed_nodes_before_softmax:
+            if not isinstance(parent, NodeProto):
+                break
+            parent_candidates = graph.get_node_parents(parent)
+            if len(parent_candidates) != 1:
+                raise ValueError(
+                    "While traversing the graph to find a Softmax that precedes "
+                    f"the candidate for a `value` MatMul: {node.name}, found a node "
+                    f"with multiple parents {parent.name}. "
+                    "It is assumed that the graph that connects the Softmax "
+                    "node and the `value` MatMul node is a linear chain of nodes "
+                    "and thus none of the encountered nodes should have multiple "
+                    "parents"
+                )
+            parent = graph.get_node_parents(parent)[0]
```

Comment:
can use `get_node_single_parent` - additionally why make another call in line 431?